Assets/_Main/Scripts/Player/PlayerHealth.cs

```diff
@@ -0,0 +1,155 @@
+    [Tooltip("Heal a heart when MaxHP should increase.")]
+    [SerializeField] private bool _healOnMaxHpUp = false;
+
+    private static PlayerHealth _instance = null;
+
+    #endregion
+
+    #region PROPERTIES
+    /// <summary> Instance of the script for singleton use. </summary>
+    public static PlayerHealth instance
+    {
+        get { return _instance; }
+    }
+    /// <summary> Upper HP limit. </summary>
+    public int maxHPLimit
+    {
+        get { return _maxHPLimit; }
+        set { _maxHPLimit = value; }
+    }
+    /// <summary> Maximum HP. </summary>
+    public int maxHealth
+    {
+        get { return _maxHealth; }
+        set { _maxHealth = value; }
+    }
+    /// <summary> Actual HP. </summary>
+    public int health
+    {
+        get { return _health; }
+        set { _health = value; }
+    }
+    /// <summary> If player recovers HP when maxHP increases. </summary>
+    public bool healOnMaxHpUp
+    {
+        get { return _healOnMaxHpUp; }
+        set { _healOnMaxHpUp = value; }
+    }
+
+    /// <summary> Delegate to call when need to update hearts. </summary>
+    public delegate void UpdateHearts();
+    /// <summary> Delegate to call when need to add hearts. </summary>
+    public delegate void AddHearts();
+
+    /// <summary> Event when player damage is taken. </summary>
+    public static event UpdateHearts OnDamageTaken;
+    /// <summary> Event when player hp is restored. </summary>
+    public static event UpdateHearts OnRestoreHP;
+    /// <summary> Event when player max hp is increased. </summary>
+    public static event AddHearts OnMaxHealthIncreased;
+
+    #endregion
+
+    #region METHODS
+    public void TakeDamage()
+    {
+        if (_health <= MINHEALTH)
+        {
+            return;
+        }
+        _health -= DAMAGETOTAKE;
+
+        if (OnDamageTaken != null)
+        {
+            OnDamageTaken.Invoke();
+        }
+    }
+
+    public void Heal()
+    {
+        if (_health >= _maxHealth)
+        {
+            return;
+        }
+        _health += LIFETORESTORE;
+
+
+        if (OnRestoreHP != null)
+        {
+            OnRestoreHP.Invoke();
+        }
+    }
+
+    public void UpgradeHealth()
+    {
+        if (_maxHealth >= _maxHPLimit)
+        {
+            if (_healOnMaxHpUp)
+            {
+                Heal();
+            }
+            return;
+        }
+        _maxHealth += LIFETOINCREASE; // "IVAN" Usar ++ o dejarlo para un objeto especial
+
+        if (_healOnMaxHpUp) // "IVAN" Se usa arriba �Hacer funci�n?
+        {
+            Heal();
+        }
+
+        if (OnMaxHealthIncreased != null)
+        {
+            OnMaxHealthIncreased.Invoke();
+        }
+    }
+
+    #endregion
+
+    #region MONOBEHAVIOUR
+
+    private void Awake()
+    {
+        if (_instance == null)
+        {
+            _instance = this;
+            return;
+        }
+        Destroy(gameObject); // "IVAN" �Es mejor destruir script?
```

Comment:
There is no need for this to be a singleton, try creating a GameManager that contains a reference to Player, and then that player contains reference to the PlayerHealth.

This means more classes but in the long run you will need this structure anyways.